### PR TITLE
feat(): add logic for calculating party halfway

### PR DESCRIPTION
**Pull request by marcusmolchany**
* add `totalAmountContributedToParyInWei` to `Contribution` type
* add api call to fetch total eth needed for party to reach goal
* add logic to calculate if party is halfway after a given contribution event

### proof for my multiply by 2 and compare logic
Screen Shot 2022-02-10 at 4 14 27 PM

https://replit.com/@marcusmolchany/AwareLopsidedSites#index.js

### screenshots of tweet in development

Screen Shot 2022-02-14 at 3 11 58 PM

### todo
- [x] decide on if api endpoint should be env var or checked into the repo
- [x] double check my big number logic for calculating gte halfway

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -50,14 +52,20 @@ export const getShouldAlertAboutPartyHalfWay = async (
...
+
+  // convert wei values to BigNumbers
+  const totalEthNeededBN = BigNumber.from(totalEthNeeded);
+  const totalContributionAmountBN = BigNumber.from(
+    event.contribution.totalAmountContributedToParyInWei
+  );
+
+  // BN.js cannot handle decimal values, so instead we multiply the total contributions by 2 and compare to
+  // total eth needed
+  const isPartyHalfway = totalContributionAmountBN.mul(2).gte(totalEthNeededBN);
```
Comment (marcusmolchany) [inline src/utils.ts:70]: is this true?

File: src/utils.ts [hunk 2]
```diff
@@ -50,14 +52,20 @@ export const getShouldAlertAboutPartyHalfWay = async (
...
 
-  // todo:
-  // 1. create api endpoint that does the following:
-  //   * if partybid - fetch auction reserve price
-  //   * if partybuy - fetch maxPrice from contract
-  //   * if collectionbuy - fetch current floor price
-  // 2. compare (fetched value) / totalContributions >= .5
-  // 3. return boolean
-  const isPartyHalfway = false;
+  const API_ENDPOINT = ""; // todo: add this
```
Comment (marcusmolchany) [inline src/utils.ts]: let me know if we should check in this api endpoint, or if it makes more sense to use an env var

**Review by marcusmolchany — COMMENTED** [review]
a couple question for @steveklebanoff @samedayjeremy

File: src/utils.ts (hunk 1 above)
Comment (steveklebanoff) [reply src/utils.ts:70]: yes bn.js doesnt support decimals

File: src/utils.ts [hunk 3]
```diff
@@ -50,14 +52,20 @@ export const getShouldAlertAboutPartyHalfWay = async (
...
 
-  // todo:
-  // 1. create api endpoint that does the following:
-  //   * if partybid - fetch auction reserve price
-  //   * if partybuy - fetch maxPrice from contract
-  //   * if collectionbuy - fetch current floor price
-  // 2. compare (fetched value) / totalContributions >= .5
-  // 3. return boolean
-  const isPartyHalfway = false;
+  const API_ENDPOINT = "https://partybid.app/api/party/price_details";
```
Comment (steveklebanoff) [inline src/utils.ts]: feels like this should be in the config even if its hard coded there

File: src/utils.ts (hunk 1 above)
Comment (steveklebanoff) [reply src/utils.ts:70]: `BN.js` doesnt but `BigNumber` does --- this is a BigNumber here so it actually can support decimal values

Comment (steveklebanoff) [reply src/utils.ts:70]: we should at least remove the comment here

**Review by steveklebanoff — APPROVED** [review]

**Pull request merged**